File: node/lib/NNet.js
import Neuron from "./neuron.js";

class NNet{
    constructor(inputSize){
        this.layers = 0
        this.step = 3.0
        this.batchSize = 10
        this.trainings = 0
        this.batches =0
        this.count = 0
        // keep a list of neurons
        // makes it easier to reset and serialize
        this.neurons = [] // Hidden and output Neurons
        this.allNeurons = []
        this.output = []
        this.input = this.addLayer(inputSize)
        this.costs = []
        this.costIndex=0
        this.costsSize = 500
    }

    addLayer(size){
        let position = 0
        let layer = Array.from({length:size}, (n1,i)=>{
            let n = new Neuron()
            n.position = position++
            n.id=this.allNeurons.length
            n.index = i
            this.allNeurons.push(n)
            this.output.forEach((other)=> n.connect(other))
            return n
        })
        this.neurons.push(...layer.filter(n=> !n.isInput())) // add all non input neurons
        this.output = layer
        this.layers++
        return layer
    }

    reset(){this.neurons.forEach((n)=>{n.reset()})}

    feed(input){
        this.count++
        this.reset()
        this.input.forEach((n,i)=> {n.value = input[i]})
        this.neurons.forEach((n, i)=>{n.ff()})
    }

    getOutput(){
        return this.output.map(n=>n.getValue())
    }

    // convenience methods
    translateInput(value){
        // allow translate of any object to list of values
        // default do nothing
        return value
    }

    translateOutput(value){
        // allow translate of list of values to object
        // default do nothing
        return value
    }

    translateExpected(value){
        // allow translate of object to expected output vector
        // default do nothing
        return value
    }

    check(item){
        this.feed(this.translateInput(item))
        return this.translateOutput(this.getOutput())
    }

    //getCost(result)
    getCost(){
        return this.output.reduce((previous,n,i)=> (previous + (((n.expected - n.value) ** 2  ) / (2*this.output.length))) , 0)
    }
    getAverageCost(length){
        // average cost of the most recent trainins
        let slice
        if(!length){
            slice = this.costs
        } else {
            slice = this.costs.slice(Math.max(0,this.costIndex-length), this.costIndex+1)
            if (this.costIndex<length) {
                slice = [...slice, ...this.costs.slice(this.costIndex - length)]
            }
        }
        return slice.reduce((previous,cost)=>{return previous+cost},0)/slice.length
    }

    getHighest(pattern){
        return pattern.reduce((previous,val,i,arr)=> val>arr[previous]?i:previous,0)
    }

    train(item, expected){
        let output = this.check(item)
        // store the expected values in the output neurons
        let expectedVector = this.translateExpected(expected)
        this.output.forEach((n,i)=> {n.expected = expectedVector[i] })
        this.costIndex = this.trainings % this.costsSize
        this.costs[this.costIndex] = this.getCost()
        this.trainings ++

        // backpropagate Delta recursive
        this.neurons.forEach((n)=>n.getDelta())

        // adjust weights
        if (this.trainings % this.batchSize == 0){
            this.batches++
            // iterative
            this.neurons.forEach((n)=>n.learn(this.step/this.batchSize))
        }
    }

    toJSON(){
        // recent cost functions or success rate
        // trainings
        // patterns?

        return {
            inputSize: this.input.length,
            outputSize: this.output.length,
            allNeurons: this.allNeurons.map(n => {
                return {
                    bias: n.bias,
                    id: n.id,
                    index: n.index,
                    connectionIDs: n.getConnectionIDs(),
                    weights: n.getWeights()
                }
            })
        }
    }
    /*printInfo(){
        console.log("counts "+counts)
    }*/

}


export default NNet

File: java/src/NNet.java
import java.util.ArrayList;
import java.util.Iterator;
import java.util.ListIterator;

public class NNet {
   private Neuron[] input;
   private Neuron[] output;
   private int layers = 0;
   public double step = 2;
   public int batchSize = 10;
   int batches = 0;
   public int trainings = 0;
   public  int count = 0;
            
    private ArrayList<Neuron> neurons;
    private double costs[] = new double[500];
    public NNet(int inputSize){
        input = addLayer(inputSize);
        output = input;
        this.neurons = new ArrayList();
    }

    public Neuron[] addLayer(int size){
        Neuron[] layer = new Neuron[size];
        for (int i = 0; i < size; i++) {
            layer[i] = new Neuron();
            if (this.neurons!=null) { // not input
                layer[i].connect(this.output);
                this.neurons.add(layer[i]);
            }
        }
        this.output = layer;
        this.layers++;
        return layer;
    }

    public void reset() {
        this.neurons.forEach(n->n.reset());
        /*for (Iterator iterator = this.neurons.iterator(); iterator.hasNext(); ) {
            ((Neuron) iterator.next()).reset();
        }*/
    }    
        
    public void feed(double[] values){
        this.count++;
        this.reset();
        for (int i = 0; i < this.input.length; i++) {
            //Neuron neuron = this.input[i];
            //neuron.setValue(values[i]);
            this.input[i].setValue(values[i]);
            /*this.input[i].value = values[i];
            this.input[i].reset = false;*/
        }
        /*this.neurons.forEach(n->n.getValue());*/
        this.neurons.forEach(n->n.ff ());
        /*Iterator<Neuron> iter = this.neurons.iterator();
        while (iter.hasNext()) {
            (iter.next()).ff();
        }*/
    }

    public double[] getOutput(){
        double[] result = new double[this.output.length];
        for (int i = 0; i < this.output.length; i++) {
            result[i]=this.output[i].getValue();
        }
        return result;
    }

    public double[] check(double[] values){
        this.feed(values);
        return this.getOutput();
    }

        //getCost(result)
    public double getCost(){
        double result = 0;
        for (int i = 0; i < this.output.length; i++) {
            Neuron n = this.output[i];
            result+=(Math.pow(n.expected - n.getValue(),2  ) / (2*this.output.length));
        }
        return result;
    }

    public int getHighest(double[] pattern){
        double high = 0;
        int highIndex =0;
        for (int i = 0; i < pattern.length; i++) {
            if(pattern[i] > high) {
                high=pattern[i];
                highIndex=i;
            }
        }
        return highIndex;
    }

    public void train(double[] values, double[] expected) {
        this.feed(values);
        // store the expected values in the output neurons
        for (int i = 0; i < this.output.length; i++) {
            this.output[i].expected = expected[i];
        }

        double[] output = this.getOutput();
        int index = this.trainings % this.costs.length;
        this.costs[index] = this.getCost();
        this.trainings++;

        /*for (int i = 0; i < this.input.length; i++) {
            Neuron neuron = this.input[i];
            neuron.getDelta();
        }*/
        Iterator iter = this.neurons.iterator();
        while (iter.hasNext()) {
            ((Neuron) iter.next()).getDelta();
        }
        /*ListIterator iter = this.neurons.listIterator(this.neurons.size());
        while (iter.hasPrevious()) {
            ((Neuron) iter.previous()).getDelta();
        }*/
        // adjust weights
        if (this.trainings % this.batchSize == 0) {
            this.batches++;
            Iterator iter2 = this.neurons.iterator();
            while (iter2.hasNext()) {
                ((Neuron) iter2.next()).learn(this.step / this.batchSize);
            }
            //System.out.println(this.getCost());
        }
    }
    public double getAverageCost(){
        double sum =0;
        for (int i = 0; i < costs.length; i++) {
            sum += costs[i];
        }
        return sum / costs.length;
    }
}
